packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -106,87 +168,229 @@ export class BlobManager {
+
+            // Blobs created while the container is detached are stored in IDetachedBlobStorage.
+            // The 'IDocumentStorageService.readBlob()' call below will retrieve these via localId.
+            storageId = blobId;
+        } else {
+            storageId = this.redirectTable.get(blobId);
+            assert(!!storageId, 0x11f /* "requesting unknown blobs" */);
+        }
 
-        // When this blob is retrieved, let the container runtime know that the corresponding GC node got updated.
+        // When a GC-able (not pending) blob is retrieved, let runtime know that the corresponding GC node got updated.
         this.gcNodeUpdated(this.getBlobGCNodePath(blobId));
 
+        return PerformanceEvent.timedExecAsync(
+            this.logger,
+            { eventName: "AttachmentReadBlob", id: storageId },
+            async () => {
+                return this.getStorage().readBlob(storageId);
+            },
+            { end: true, cancel: "error" },
+        );
+    }
+
+    private getBlobHandle(id: string): IFluidHandle<ArrayBufferLike> {
+        assert(this.redirectTable.has(id) || this.pendingBlobs.has(id),
+            "requesting handle for unknown blob");
         return new BlobHandle(
-            `${BlobManager.basePath}/${storageId}`,
+            `${BlobManager.basePath}/${id}`,
             this.routeContext,
-            async () => PerformanceEvent.timedExecAsync(
-                this.logger,
-                { eventName: "AttachmentReadBlob", id: storageId },
-                async () => {
-                    return this.getStorage().readBlob(storageId);
-                },
-                { end: true, cancel: "error" },
-            ),
+            async () => this.getBlob(id),
         );
     }
 
+    private async createBlobDetached(blob: ArrayBufferLike): Promise<IFluidHandle<ArrayBufferLike>> {
+        // Blobs created while the container is detached are stored in IDetachedBlobStorage.
+        // The 'IDocumentStorageService.createBlob()' call below will respond with a localId.
+        const response = await this.getStorage().createBlob(blob);
+        this.redirectTable.set(response.id, undefined);
+        return this.getBlobHandle(response.id);
+    }
+
     public async createBlob(blob: ArrayBufferLike): Promise<IFluidHandle<ArrayBufferLike>> {
+        if (this.runtime.attachState === AttachState.Detached) {
+            return this.createBlobDetached(blob);
+        }
         if (this.runtime.attachState === AttachState.Attaching) {
             // blob upload is not supported in "Attaching" state
             this.logger.sendTelemetryEvent({ eventName: "CreateBlobWhileAttaching" });
             await new Promise<void>((resolve) => this.runtime.once("attached", resolve));
         }
+        assert(this.runtime.attachState === AttachState.Attached,
+            "For clarity and paranoid defense against adding future attachment states");
+
+        const localId = uuid();
```

Comment:
If it helps at all, you could construct local ids in a way that distinguishes them from storage ids, so you can tell which one you have without having search (just an idea).